language/move-lang/src/typing/infinite_instantiations.rs

```diff
@@ -0,0 +1,423 @@
+    use T::SequenceItem_ as S;
+    match &item.value {
+        S::Bind(_, _, te) | S::Seq(te) => exp(context, te),
+        S::Declare(_) => (),
+    }
+}
+
+fn exp(context: &mut Context, e: &T::Exp) {
+    use T::UnannotatedExp_ as E;
+    match &e.exp.value {
+        E::InferredNum(_) | E::Use(_) => panic!("ICE should have been expanded"),
+
+        E::Unit
+        | E::Value(_)
+        | E::Move { .. }
+        | E::Copy { .. }
+        | E::BorrowLocal(_, _)
+        | E::Break
+        | E::Continue
+        | E::Spec(_, _)
+        | E::UnresolvedError => (),
+
+        E::ModuleCall(call) => {
+            context.add_usage(e.exp.loc, &call.module, &call.name, &call.type_arguments);
+            exp(context, &call.arguments)
+        }
+
+        E::IfElse(eb, et, ef) => {
+            exp(context, eb);
+            exp(context, et);
+            exp(context, ef);
+        }
+        E::While(eb, eloop) => {
+            exp(context, eb);
+            exp(context, eloop);
+        }
+        E::Loop { body: eloop, .. } => exp(context, eloop),
+        E::Block(seq) => sequence(context, seq),
+        E::Assign(_, _, er) => exp(context, er),
+
+        E::Builtin(_, er)
+        | E::Return(er)
+        | E::Abort(er)
+        | E::Dereference(er)
+        | E::UnaryExp(_, er)
+        | E::Borrow(_, er, _)
+        | E::TempBorrow(_, er) => exp(context, er),
+        E::Mutate(el, er) | E::BinopExp(el, _, _, er) => {
+            exp(context, el);
+            exp(context, er)
+        }
+
+        E::Pack(_, _, _, fields) => {
+            for (_, (_, (_, fe))) in fields.iter() {
+                exp(context, fe)
+            }
+        }
+        E::ExpList(el) => exp_list(context, el),
+
+        E::Cast(e, _) | E::Annotate(e, _) => exp(context, e),
+    }
+}
+
+fn exp_list(context: &mut Context, items: &[T::ExpListItem]) {
+    items.iter().for_each(|item| exp_list_item(context, item))
+}
+
+fn exp_list_item(context: &mut Context, item: &T::ExpListItem) {
+    use T::ExpListItem as I;
+    match item {
+        I::Single(e, _) | I::Splat(_, e, _) => {
+            exp(context, e);
+        }
+    }
+}
+
+//**************************************************************************************************
+// Errors
+//**************************************************************************************************
+
+fn cycle_error(context: &Context, graph: &DiGraphMap<&TParam, Edge>, scc: Vec<&TParam>) -> Error {
+    let critical_edge = scc_edges!(graph, &scc).find(|(_, e, _)| e == &Edge::Nested);
+    // tail -> head
+    let (critical_tail, _, critical_head) = critical_edge.unwrap();
+    let (_, cycle_nodes) = petgraph_astar(
+        graph,
+        critical_head,
+        |finish| &finish == critical_tail,
+        |_e| 1,
+        |_| 0,
+    )
+    .unwrap();
+    let next = |i| {
+        if i + 1 == cycle_nodes.len() {
+            0
+        } else {
+            i + 1
+        }
+    };
```

Comment:
Why not do `|i| { (i + 1) % cycle_nodes.len()}` for this?

Reply:
Fancy. 

I sort of forgot to simplify this (it used to return the element instead of the index)